pkg/apis/seedmanagement/validation/managedseed.go

```diff
@@ -0,0 +1,126 @@
+func ValidateManagedSeedSpec(spec *seedmanagement.ManagedSeedSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	// Check that either seedTemplate or gardenlet is specified
+	if (spec.SeedTemplate == nil && spec.Gardenlet == nil) || (spec.SeedTemplate != nil && spec.Gardenlet != nil) {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("seedTemplate/gardenlet"), spec, "exactly one of `seedTemplate` and `gardenlet` is required"))
+	}
+
+	switch {
+	case spec.SeedTemplate != nil:
+		seedTemplatePath := fldPath.Child("seedTemplate")
+
+		// Validate seed template
+		allErrs = append(allErrs, metav1validation.ValidateLabels(spec.SeedTemplate.Labels, seedTemplatePath.Child("metadata", "labels"))...)
+		allErrs = append(allErrs, apivalidation.ValidateAnnotations(spec.SeedTemplate.Annotations, seedTemplatePath.Child("metadata", "annotations"))...)
+		allErrs = append(allErrs, corevalidation.ValidateSeedSpec(&spec.SeedTemplate.Spec, seedTemplatePath.Child("spec"))...)
+
+	case spec.Gardenlet != nil:
+		gardenletPath := fldPath.Child("gardenlet")
+		gardenletConfigPath := gardenletPath.Child("config")
+		seedConfigPath := gardenletConfigPath.Child("seedConfig")
+
+		// Ensure gardenlet config is specified
+		if spec.Gardenlet.Config == nil {
+			allErrs = append(allErrs, field.Required(gardenletConfigPath, "gardenlet config is required"))
+		}
+
+		// Decode gardenlet config from raw extension
+		gardenletConfig, err := helper.DecodeGardenletConfig(spec.Gardenlet.Config)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(gardenletConfigPath, spec.Gardenlet.Config, fmt.Sprintf("could not decode gardenlet config: %v", err)))
+		}
+
+		// Validate gardenlet config
+		allErrs = append(allErrs, configvalidation.ValidateGardenletConfiguration(gardenletConfig)...)
+
+		// Ensure seed config is specified
+		if gardenletConfig.SeedConfig == nil {
+			allErrs = append(allErrs, field.Required(seedConfigPath, "seed config is required"))
+		}
+
+		// Ensure seed name is empty since it will be set by the controller
+		if gardenletConfig.SeedConfig.Name != "" {
+			allErrs = append(allErrs, field.Forbidden(seedConfigPath.Child("metadata", "name"), fmt.Sprintf("seed name must not be specified")))
+		}
+
+		// Convert seed config to internal version
+		seed, err := helper.ConvertSeed(&gardenletConfig.SeedConfig.Seed)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(seedConfigPath, gardenletConfig.SeedConfig.Seed, fmt.Sprintf("could not convert seed config: %v", err)))
+		}
+
+		// Validate seed config
+		allErrs = append(allErrs, metav1validation.ValidateLabels(gardenletConfig.SeedConfig.Labels, seedConfigPath.Child("metadata", "labels"))...)
+		allErrs = append(allErrs, apivalidation.ValidateAnnotations(gardenletConfig.SeedConfig.Annotations, seedConfigPath.Child("metadata", "annotations"))...)
+		allErrs = append(allErrs, corevalidation.ValidateSeedSpec(&seed.Spec, seedConfigPath.Child("spec"))...)
```

Comment:
Yes, I had the same thoughts, I am glad that you propose it as well. I would have done it this way but wasn't sure of the consequences. Still, perhaps we should keep it as it is for now and change it later, to avoid some unforseen consequences.

Reply:
I don't see any concern with moving it to the `GardenletConfiguration` as of now. Though, no guarantee, it will work 😄.
The gardenlet config is anyway a bit too loose from my POV.
Also, I just noticed, the config API includes a complete Seed object (including kind/version and status) instead of a template composed of metadata and spec or a `RawExtensions`.
So I guess, this needs to be reworked a bit.

Considering this, I'm fine with both doing it before this PR and using it, or doing it later on and move the validation code here then.

Reply:
OK, let's keep it like it is then and rework it later along the lines you suggest, to avoid making this PR even bigger.